Add support for vim-clap

meck opened the pull request:
This adds highlights for [vim-clap](https://github.com/liuchengxu/vim-clap)

<img width="1068" alt="clap" src="https://user-images.githubusercontent.com/762115/69010198-984ccd00-095d-11ea-956c-01d7aa9d832a.png">

arcticicestudio commented:
Hi @meck :wave:, thanks for your contribution :+1:
I'll review it in the near future and get back to you.

Arelav commented:
Can this be merged, please?

sdemura commented:
I would love to see this merged!

ikalnytskyi commented on colors/nord.vim:
```diff
@@ -601,6 +601,49 @@ hi! link NERDTreeHelp Comment
+call s:hi("ClapSelected", s:nord7_gui, "", s:nord7_term, "", s:underline, "")
+call s:hi("ClapCurrentSelection", s:nord7_gui, "", s:nord8_term, "", s:bold, "")
+call s:hi("ClapNoMatchesFound", s:nord13_gui, "", s:nord13_term, "", "", "")
+let clap_matches = [
+        \ [s:nord11_gui, s:nord11_term] ,
+        \ [s:nord13_gui, s:nord13_term] ,
+        \ [s:nord14_gui, s:nord14_term] ,
+        \ [s:nord11_gui, s:nord11_term] ,
+        \ [s:nord13_gui, s:nord13_term] ,
+        \ [s:nord14_gui, s:nord14_term] ,
+        \ [s:nord11_gui, s:nord11_term] ,
+        \ [s:nord13_gui, s:nord13_term] ,
+        \ ]
+let idx = 1
+for g in clap_matches
+  call s:hi("ClapMatches" . idx, g[0], "", g[1], "", "", "")
+  let idx += 1
+endfor
+let clap_fuzzy_matches = [
+        \ [s:nord11_gui, s:nord11_term] ,
+        \ [s:nord13_gui, s:nord13_term] ,
+        \ [s:nord14_gui, s:nord14_term] ,
+        \ [s:nord11_gui, s:nord11_term] ,
+        \ [s:nord13_gui, s:nord13_term] ,
+        \ [s:nord14_gui, s:nord14_term] ,
+        \ [s:nord11_gui, s:nord11_term] ,
+        \ [s:nord13_gui, s:nord13_term] ,
+        \ [s:nord14_gui, s:nord14_term] ,
+        \ [s:nord11_gui, s:nord11_term] ,
+        \ [s:nord13_gui, s:nord13_term] ,
+        \ [s:nord14_gui, s:nord14_term] ,
+        \ ]
+let idx = 1
+for g in clap_fuzzy_matches
+  call s:hi("ClapFuzzyMatches" . idx, g[0], "", g[1], "", "", "")
+  let idx += 1
+endfor
```
Good work! I'm not maintainer of this repo, but may I suggest a slight improvement over this code?

It seems `clap_matches` and `clap_fuzzy_matches` are pretty much the same (the only difference is length). In fact, they are only 3 colors that we want repeat over and over again. So my suggestion is simply maintain a table of 3 colors, and pick the right one based on the index of `ClapMatch` we're creating.

The only side effect is that we will produce more highlight rules for `ClapMatch` than the plugin supports (12 instead of 8), but in my opinion it's not a big deal. What do you think?


```suggestion
let clap_matches = [
        \ [s:nord11_gui, s:nord11_term] ,
        \ [s:nord13_gui, s:nord13_term] ,
        \ [s:nord14_gui, s:nord14_term] ,
        \ ]
let idx = 1
while idx <= 12
  let clap_match_color = clap_matches[idx % len(clap_matches) - 1]
  call s:hi("ClapMatches" . idx, clap_match_color[0], "", clap_match_color[1], "", "", "")
  call s:hi("ClapFuzzyMatches" . idx, clap_match_color[0], "", clap_match_color[1], "", "", "")
  let idx += 1
endfor
````

meck replied:
LGTM @arcticicestudio can decide

arcticicestudio commented:
Sorry for this delay, life, other projects and tons of _Nord_ issues/PRs keeps me busy 😐
I haven't checked and tested the actual code, but I'd like to throw in some thoughts regarding the design.
_Nord_'s main accent is `nord8` that can be combined with the other colors of the _Frost_ palette. While _Aurora_ colors like `nord12` or `nord13` might be used more often in other themes, _Nord_ tries to keep they purpose as _signal colors_ to style e.g. warnings and errors.
Therefore I'd like to propose to change the style for _Vim Clap_ to use

- `nord4` as default foreground color.
- `nord8` as main color for matches
- `nord9` and `nord10` for sub-matches since they have a darker appearance and `nord8` should get the user focus while these both indicate that it might also match. 

I use [fzf][] as my main _fuzzy finder_ for almost everything so I also use it together with [fzf.vim][]. While there's no official _Nord_ port project for it (yet) I created a _Nord_ theme for my personal use that is embedded into _.dotfiles_ (large ZSH update coming soon™️).
This is what is looks like when using the design I've described above:

<p align="center"><img src="https://user-images.githubusercontent.com/7836623/80416737-ebe02100-88d4-11ea-9a5c-986b2ad7347f.png" /></p>

Thanks again for the active feedback on this PR, really like the way the community works on such things 😄 
Let me know what you think if we'd use this design for _Vim Clap_ too.

[fzf]: https://github.com/junegunn/fzf
[fzd.vim]: https://github.com/junegunn/fzf.vim

meck commented:
I've cleaned this up a bit and conformed it to @arcticicestudio's suggestions, however I noticed Clap now has a auto-loaded theme system as used in liuchengxu/vim-clap#420. So it should probably be moved there (The same as airline and sightline)? Also all the providers have their own hl-groups, I've added a couple but the rest should probably be added as well.

Then there is the question if there should be a theme file in here and in the clap repo as well? I guess that is something for @liuchengxu and @arcticicestudio to agree on.

liuchengxu commented:
>Then there is the question if there should be a theme file in here and in the clap repo as well?

The advantage of putting it in the colorscheme repo is when people use this colorscheme, they'll have the clap support aumatically and don't have to add the line `let g:clap_theme = 'nord'` explicitly. So I would second the former way if the clap theme has a original vim colorscheme.

arcticicestudio commented on colors/nord.vim:
```diff
@@ -572,7 +572,9 @@ call s:hi("LSPDiagnosticsWarning", s:nord13_gui, "", s:nord13_term, "", "", "")
 call s:hi("LSPDiagnosticsInformation" , s:nord8_gui, "", s:nord8_term, "", "", "")
 call s:hi("LSPDiagnosticsHint" , s:nord10_gui, "", s:nord10_term, "", "", "")
-
+hi! link LSPReferenceText CursorColumn
+hi! link LSPReferenceRead LSPReferenceText
+hi! link LSPReferenceWrite LSPReferenceText
```
I guess this was added because this PR is created from the default (`develop`) branch instead of a story branch so changed to the fork are reflected here. Should be removed and maybe added in a new PR when it was intended to be included.

```suggestion
```

arcticicestudio commented:
---

###### Release Notes Assets

<p align="center"><img src="https://user-images.githubusercontent.com/7836623/86534351-0daee300-bed8-11ea-9685-031d16b5a07b.png" /></p>
